sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1WriteFn.java

```diff
@@ -0,0 +1,526 @@
+   */
+  static abstract class BaseBatchWriteFn<Out> extends WindowAwareDoFn<Write, Out> implements
+      HasRpcAttemptContext {
+    private static final Logger LOGGER = LoggerFactory.getLogger(FirestoreV1Fn.V1FnRpcAttemptContext.BatchWrite.getNamespace());
+    private final JodaClock clock;
+    private final FirestoreStatefulComponentFactory firestoreStatefulComponentFactory;
+    private final RpcQosOptions rpcQosOptions;
+
+    // transient running state information, not important to any possible checkpointing
+    private transient FirestoreRpc firestoreRpc;
+    private transient RpcQos rpcQos;
+    private transient String projectId;
+    @VisibleForTesting
+    transient Queue<@NonNull WriteElement> writes = new PriorityQueue<>(WriteElement.COMPARATOR);
+    @VisibleForTesting
+    transient int queueNextEntryPriority = 0;
+
+    @SuppressWarnings("initialization.fields.uninitialized") // allow transient fields to be managed by component lifecycle
+    protected BaseBatchWriteFn(
+        JodaClock clock,
+        FirestoreStatefulComponentFactory firestoreStatefulComponentFactory,
+        RpcQosOptions rpcQosOptions
+    ) {
+      this.clock = clock;
+      this.firestoreStatefulComponentFactory = firestoreStatefulComponentFactory;
+      this.rpcQosOptions = rpcQosOptions;
+    }
+
+    protected Logger getLogger() {
+      return LOGGER;
+    }
+
+    @Override
+    public Context getRpcAttemptContext() {
+      return FirestoreV1Fn.V1FnRpcAttemptContext.BatchWrite;
+    }
+
+    @Override
+    public final void populateDisplayData(@edu.umd.cs.findbugs.annotations.NonNull DisplayData.Builder builder) {
+      builder
+          .include("rpcQosOptions", rpcQosOptions);
+    }
+
+    @Override
+    public void setup() {
+      rpcQos = firestoreStatefulComponentFactory.getRpcQos(rpcQosOptions);
+      writes = new PriorityQueue<>(WriteElement.COMPARATOR);
+    }
+
+    @Override
+    public final void startBundle(StartBundleContext c) {
+      String project = c.getPipelineOptions().as(GcpOptions.class).getProject();
+      projectId = requireNonNull(project, "project must be defined on GcpOptions of PipelineOptions");
+      firestoreRpc = firestoreStatefulComponentFactory.getFirestoreRpc(c.getPipelineOptions());
+    }
+
+    /**
+     * For each element extract and enqueue all writes from the commit. Then potentially flush any
+     * previously and currently enqueued writes.
+     * <p/>
+     * In order for writes to be enqueued the value of {@link BatchWriteRequest#getDatabase()} must
+     * match exactly with the database name this instance is configured for via the provided {@link
+     * org.apache.beam.sdk.options.PipelineOptions PipelineOptions}
+     * <p/>
+     * {@inheritDoc}
+     */
+    @Override
+    public void processElement(ProcessContext context, BoundedWindow window) throws Exception {
+      @SuppressWarnings("nullness") // for some reason requireNonNull thinks its parameter must be non-null...
```

Comment:
I spent a little time digging and it appears to be due to the fact that error checker is configured to assume non null unless explicitly annotated, so `Objects.requireNotNull` is being treated as only accepting NonNull arguments. I'll update the comment to state this instead of what's there right now.